### ISSUE TITLE
Problems with VRAM after the last update.

### ISSUE THREAD

**Issue by okolenmi** [comfyanonymous/ComfyUI#1332]
After the last update I have a lot of problems with VRAM exhaustion even on 512*768 images. Default VAE Decoder consumes less resources than tiled one. Also default decoder works worse than tiled from previous versions, as I can see on VRAM consuption graph...
There is also another problem. Now I have the same issue as in A1111's UI: after a few generations I have not enough VRAM to generate anything and it chrashes on VAE Decode step. 

Torch 1.13.1+cu117. Also tried it on recomended in installation guide cu113 - the same result.
**No extensions.** (I have tried a new installation)

Default img2img 768*512 generation with 2GB checkpoint only (no lora) consumes 3GB+ of VRAM(0.4GB were used by Windows10) 
Red line is a point where VAE Decode (Tiled) starts to work and then UI crash happens after memory flow to shared section. 
![изображение](https://github.com/comfyanonymous/ComfyUI/assets/55319814/756d5fd7-f0aa-4983-b745-8c08a1aabd4d)

**Comment by GalaxyTimeMachine** [comment]
I'm seeing similar VRAM usage issues with a 4090 (24Gb). I am using SDXL base and refiner and then tiled upscaling and face restore with a SD1.5 model. The VRAM is all used, and it's using shared VRAM (which is VERY slow). It's currently stuck trying to create a single 832x1216 image and showing 126s/it! I used to be able to create a batch of 2 and the VRAM was hardly used.
![image](https://github.com/comfyanonymous/ComfyUI/assets/52193044/3a4da4ed-305d-42c4-8618-4ecb5011b33c)

**Comment by GalaxyTimeMachine** [comment]
I've just rolled back to an older Nvidia driver to see if it helps. Are models being loaded into RAM and VRAM at the same time? In this image, you can see the step-up as each model is being loaded in both VRAM & RAM. This was captured after the image had finished being created, and both of them are still being utilised by something.
![image](https://github.com/comfyanonymous/ComfyUI/assets/52193044/756e59ab-fa51-40d0-b6ee-855bdd98e6c6)

**Comment by GalaxyTimeMachine** [comment]
The older Nvidia driver is a bit better, but my single image creation times are double what they were a day or two ago. I never used to get these messages, but do on every gen now:
![image](https://github.com/comfyanonymous/ComfyUI/assets/52193044/2734e69a-1a6d-49cf-8bde-2901bc960398)

**Comment by GalaxyTimeMachine** [comment]
I rolled back to 30eb92c3cbe0e0dfa442d452b5f1187c654e572e and thought there was an improvement. Each image is being created about 20-30 seconds faster now, but after 3 or 4 images, the VRAM usage starts using shared VRAM, and it slows it down again. There are no model changes in between the images.
This is the VRAM at the start and after creating 1 image:
![image](https://github.com/comfyanonymous/ComfyUI/assets/52193044/31394052-bccd-4413-909c-06598b143381)
This is after creating 5 images (it isn't running anything at this stage):
![image](https://github.com/comfyanonymous/ComfyUI/assets/52193044/6417d38b-33cc-4b49-917f-73cce73b1d7c)

**Comment by eldragon0** [comment]
Just tested with .\python_embeded\python.exe -s ComfyUI\main.py --lowvram --windows-standalone-build  
low vram tag appears to work as a workaround , all of my memory issues every gen pushes me up to about 16 GB vram and after the gen it drops back down to 12.8.

**Comment by comfyanonymous** [comment]
Can you download the latest standalone don't install any custom nodes and try to reproduce the issue on it with? If you do give the the workflow.

**Comment by GalaxyTimeMachine** [comment]
Is it expected that the VRAM will be held after the image is created? Like this:
![image](https://github.com/comfyanonymous/ComfyUI/assets/52193044/ccf1b2f8-bd7a-4e0b-b8c3-7ece7def261d)
I'm not going to be able to recreate the whole problem with such a simple workflow, it only does half of what mine does, but I used the one in this image:
![ComfyUI_00002_](https://github.com/comfyanonymous/ComfyUI/assets/52193044/6492ed13-830e-40de-90eb-b0b71695570d)

**Comment by comfyanonymous** [comment]
yes it's expected that it keeps weights in GPU memory when you have enough of it. What's not expected is that it uses all of it and OOMs.

By asking you to reproduce the issue on base only I'm checking if it's a bug with the base code or if it's a custom node that has a memory leak issue.

**Comment by GalaxyTimeMachine** [comment]
It's going to be difficult to reproduce with just base nodes, because I was using Ultimate SD Upscaler (which loads other models) and also face detector/upscaler. I've been using the same workflow for some time and not had the issue, but the ttN pipe loader and ksampler did update today. Do you have any base node workflows that may stress the system a bit more?

**Comment by GalaxyTimeMachine** [comment]
I just added the custom nodes to the freshly installed windows_portable and used my workflow. After the first image gen, the used VRAM was 19.0GB (at idle), after the next it was 19.7, then 20.9
the increase happens when the pipeKSamplerSDXL is running, which then makes the Ultimate SD Upscaler and face detailer slow, because it's using shared GPU memory (disk).
While I was running the next one, I got some errors that may be more useful (?). Does this point to the tinyterrorNodes?

```
making attention of type 'vanilla-xformers' with 512 in_channels
building MemoryEfficientAttnBlock with 512 in_channels...
loading new
loading new
100%|##################################################################################| 20/20 [00:02<00:00,  7.27it/s]
loading new
100%|##################################################################################| 20/20 [00:02<00:00,  7.27it/s]
Warning: Ran out of memory when regular VAE decoding, retrying with tiled VAE decoding.
!!! Exception during processing !!!
Traceback (most recent call last):
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\sd.py", line 669, in decode
    pixel_samples[x:x+batch_number] = torch.clamp((self.first_stage_model.decode(samples) + 1.0) / 2.0, min=0.0, max=1.0).cpu().float()
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\ldm\models\autoencoder.py", line 94, in decode
    dec = self.decoder(z)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\ldm\modules\diffusionmodules\model.py", line 727, in forward
    h = self.up[i_level].upsample(h)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\ldm\modules\diffusionmodules\model.py", line 61, in forward
    x = self.conv(x)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\conv.py", line 463, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\conv.py", line 459, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
torch.cuda.OutOfMemoryError: Allocation on device 0 would exceed allowed memory. (out of memory)
Currently allocated     : 18.59 GiB
Requested               : 504.00 MiB
Device limit            : 23.99 GiB
Free (according to CUDA): 0 bytes
PyTorch limit (set by user-supplied memory fraction)
                        : 17179869184.00 GiB

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "F:\ComfyUI_windows_portable\ComfyUI\execution.py", line 151, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
  File "F:\ComfyUI_windows_portable\ComfyUI\execution.py", line 81, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
  File "F:\ComfyUI_windows_portable\ComfyUI\execution.py", line 74, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
  File "F:\ComfyUI_windows_portable\ComfyUI\custom_nodes\ComfyUI_tinyterraNodes\tinyterraNodes.py", line 1654, in sample    return process_sample_state(sdxl_pipe, base_steps, refiner_steps, cfg, sampler_name, scheduler, denoise, image_output, save_prefix, prompt, my_unique_id, preview_latent)
  File "F:\ComfyUI_windows_portable\ComfyUI\custom_nodes\ComfyUI_tinyterraNodes\tinyterraNodes.py", line 1618, in process_sample_state
    sdxl_pipe["images"] = sdxl_pipe["refiner_vae"].decode(latent).cpu()
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\sd.py", line 672, in decode
    pixel_samples = self.decode_tiled_(samples_in)
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\sd.py", line 638, in decode_tiled_
    (utils.tiled_scale(samples, decode_fn, tile_x // 2, tile_y * 2, overlap, upscale_amount = 8, pbar = pbar) +
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\utils.py", line 342, in tiled_scale
    ps = function(s_in).cpu()
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\sd.py", line 636, in <lambda>
    decode_fn = lambda a: (self.first_stage_model.decode(a.to(self.vae_dtype).to(self.device)) + 1.0).float()
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\ldm\models\autoencoder.py", line 94, in decode
    dec = self.decoder(z)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\ldm\modules\diffusionmodules\model.py", line 727, in forward
    h = self.up[i_level].upsample(h)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "F:\ComfyUI_windows_portable\ComfyUI\comfy\ldm\modules\diffusionmodules\model.py", line 61, in forward
    x = self.conv(x)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\conv.py", line 463, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "F:\ComfyUI_windows_portable\python_embeded\lib\site-packages\torch\nn\modules\conv.py", line 459, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
torch.cuda.OutOfMemoryError: Allocation on device 0 would exceed allowed memory. (out of memory)
Currently allocated     : 18.91 GiB
Requested               : 256.00 MiB
Device limit            : 23.99 GiB
Free (according to CUDA): 0 bytes
PyTorch limit (set by user-supplied memory fraction)
                        : 17179869184.00 GiB

Prompt executed in 27.29 seconds
```

**Comment by okolenmi** [comment]
### Another "route" of this discussion.
I have downloaded [the last build](https://github.com/comfyanonymous/ComfyUI/releases/tag/latest) + unistalled 1.13.1 torch and installed 2.0.1 instead.

It looks like I have a 25% better processing speed on my 4GB GTX 970 with torch 2.0.1 (+cu118). Also I have no problems with memory exhaustion after around 10 generations. Default decoder now have no problems, but there is still strange things with tiled and default VAE Decoders.

Tiled VAE Decode can be a reason of blue screen in some cases (it happens one time...). 
Unsuccessful run looks like this. A sharp rise before an emergency stop - the start of the decoder.
![Tiled](https://github.com/comfyanonymous/ComfyUI/assets/55319814/b50c1650-814f-4e1a-9110-00bd2ef76f4c)

It worked on this simple workflow (inside of the image):
![ComfyUI_00595_](https://github.com/comfyanonymous/ComfyUI/assets/55319814/9c607151-3b08-4bea-a829-721ac65bc543)

Also tiled have no clear memory feature, when default cleared it completely (can it be optional checkbox?)
Successful run with Tiled VAE Decode:
![изображение](https://github.com/comfyanonymous/ComfyUI/assets/55319814/abc09b1e-8347-48a8-9eb3-cf0e68a85559)

Successful run with Default VAE Decode
![изображение](https://github.com/comfyanonymous/ComfyUI/assets/55319814/fd8b982d-c8d6-4e8a-8190-ec3860789f82)

**So, it looks like torch 2.0.1 is the most stable thing for latest builds. Or in the pre-release build, there were some changes that positively affected the work.**
_To install pytorch 2.0.1 it's necessarily to uninstall 1.13.1 first._

**Comment by GalaxyTimeMachine** [comment]
@comfyanonymous I think I've determined that it is the ttN custom nodes causing my problem. I replaced them with your example flow and just bolted my upscale/detailer onto the end of it. The VRAM usage never exceeds 20GB, and it then drops to 14.6GB during idle. Only 20 secs to create an SDXL image now!
![image](https://github.com/comfyanonymous/ComfyUI/assets/52193044/92eafa58-70a5-4bc3-b348-046e4be59866)
![image](https://github.com/comfyanonymous/ComfyUI/assets/52193044/cd2ce5c6-99af-477d-af00-570bd134a3e1)
@okolenmi Sorry for invading your ticket space, but sounds like you made progress too.
